osu.Game/Online/Spectator/SpectatorStreamingClient.cs

```diff
@@ -170,28 +113,24 @@ async Task tryUntilConnected()
                             watchingUsers.Clear();
                         }
 
-                        // success
-                        isConnected = true;
-
-                        // resubscribe to watched users
+                        // resubscribe to watched users.
                         foreach (var userId in users)
                             WatchUser(userId);
 
                         // re-send state in case it wasn't received
                         if (isPlaying)
                             beginPlaying();
-
-                        break;
                     }
-                    catch (Exception e)
+                    else
                     {
-                        Logger.Log($"Spectator client connection error: {e}", LoggingTarget.Network);
-                        await Task.Delay(5000);
+                        playingUsers.Clear();
                     }
-                }
+                }, true);
             }
         }
 
+        protected virtual HubClientConnector CreateConnector(string name, string endpoint, IAPIProvider api) => new HubClientConnector(name, endpoint, api);
```

Comment:
really dislike this being here just for tests. would prefer a bool at this point `ShouldActuallyConnect` or something explicit. the whole nullability of this feels flimsy.